feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,188 @@
+		NotifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+	urlDef, err := url.ParseRequestURI(config.NotifyApiRoot)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing notify api root - %s", err)
+	}
+	basePath := fmt.Sprintf("%s/:%s", urlDef.Path, EncodedSessionId)
+	// Register the handlers
+	handler.NotifyServer.POST(basePath+"/update", handler.PostSmPolicyUpdateNotification)
+	handler.NotifyServer.POST(basePath+"/terminate", handler.PostSmPolicyTerminateNotification)
+
+	return handler, nil
+}
+
+func (handler *NotificationHandler) Start() (string, error) {
+	errChan := make(chan error)
+	addr, err := net.ResolveTCPAddr("tcp", handler.Config.LocalAddr)
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		err = handler.NotifyServer.Start(addr.String())
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	return sbi.WaitForSbiServerAndGetAddr(handler.NotifyServer, errChan)
+}
+
+func (handler *NotificationHandler) Shutdown() {
+	ctx, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
+	defer cancel()
+	handler.NotifyServer.Shutdown(ctx)
+}
+
+func (handler *NotificationHandler) PostSmPolicyUpdateNotification(ctx echo.Context) error {
+	client, err := relay.GetSessionProxyResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	defer client.Close()
+
+	var smUpdateNotify n7_client.SmPolicyNotification
+	err = ctx.Bind(&smUpdateNotify)
+	if err != nil {
+		glog.Errorf("Invalid SmPolicyNotification received: %s", err)
+		return ctx.NoContent(http.StatusBadRequest)
+	}
+	smPolicyDecision := smUpdateNotify.SmPolicyDecision
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for UpdateNotify- %s", err)
+		return err
+	}
+
+	policyReauthProto := n7.GetPolicyReauthRequestProto(sessionId, imsi, smPolicyDecision)
+	ans, err := client.PolicyReAuth(context.Background(), policyReauthProto)
+	if err != nil {
+		glog.Errorf("Error relaying N7 policy reauth request to gateway: %s", err)
+		return err
+	}
+	partialSuccessReport := n7.GetPartialSuccessReportN7(ans)
+	if partialSuccessReport == nil {
+		return ctx.NoContent(http.StatusOK)
+	}
+	return ctx.JSON(http.StatusOK, partialSuccessReport)
+}
+
+func (handler *NotificationHandler) PostSmPolicyTerminateNotification(ctx echo.Context) error {
+	client, err := relay.GetAbortSessionResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	defer client.Close()
+
+	var smTermNotify n7_client.TerminationNotification
+	err = ctx.Bind(&smTermNotify)
+	if err != nil {
+		glog.Errorf("Invalid SmPolicyNotification received: %s", err)
+		return ctx.NoContent(http.StatusBadRequest)
+	}
+
+	// smPolicyUrl := string(smTermNotify.ResourceUri)
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for TerminateNotify - %s", err)
+		return err
```

Comment:
here is where you should convert your error into http code (unles the reason is different for each error this function returns)